Client authentication with JWT assertion

rlewczuk opened the pull request:
Ready for preliminary review, yet still there are several things to update plus a bunch of questions.

Question: I've factored out PKCE verification part from OAuth2ClientAuthenticationProviderTests and reused it in client assertion authentication provider. Should it be merged as part of this change or separately ? 

Question: decision on how to map (decoded and verified) client assertion onto registered client is delegated to RegisterClientResolver interface. While there is default (trivial) implementation, I'm wondering if it is 'battery included' solution ? Should be extended somehow ? 

For now PR consists of several commits, will be merged into a single big commit after review+corrections cycles.

Issue gh-59

Note: new client authentication methods constants will need to be aligned with what was implemented in spring-security oauth2 client recently (v. 5.5).

jgrandja commented:
Thanks for the PR @rlewczuk !

Just a heads up that I won't be able to review this for another couple of weeks or so.
I have a few higher priority tasks that need to be completed first.

At first glance in the PR, we will need to reduce the amount of public API's exposed.
As well, this PR will need to be rebased after I add the capability of customizing client authentication methods.

I'll let you know when things are ready on my end and then we can go from there.
Thanks!

rlewczuk commented:
I've updated PR, so now PKCE verification is hidden from client code. Not sure if there is more that can be remove from public APIs at the moment. Yet, there are two issues: adapting to new customization API (#139), implementing `client_secret_jwt` and `private_key_jwt` instead of currently implemented one generic jwt client authentication method (with batteries not included). So @jgrandja, for now this PR is probably not worth your time when it comes to reviewing, I'm waiting until customization parts settle down.

jgrandja commented:
Hi @rlewczuk. I'm finally circling back to this PR as we're looking to add this feature in `0.2.1`. Are you still available to work on this?

rlewczuk commented:
Hi,

Sorry for not reacting on it, I was involved in significant project
recently.

I'll try to look at it over weekend and get back with some code next week.

Regards,
rle

On Thu, Sep 23, 2021 at 9:45 PM Joe Grandja ***@***.***>
wrote:

> Hi @rlewczuk <https://github.com/rlewczuk>. I'm finally circling back to
> this PR as we're looking to add this feature in 0.2.1. Are you still
> available to work on this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spring-projects/spring-authorization-server/pull/293#issuecomment-926103590>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMPMH4ERCLWJFEBUZ2XZEDUDN7VFANCNFSM4466ISJA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

rlewczuk-softpos commented:
@jgrandja  I've implemented converter and provider parts of this in a way that full client assertion verification (signature and against registered client repository) is delegated to a specialized JwtDecoder instance. This keeps authentication provider minimal and delegates most of work to decoder. Decoder which will validate assertion against registered client repository still needs to be implemented but I'm not sure this is valid way of thinking, So I'm stopping here for a moment and await for feedback. Can I ask for preliminary review and green light to go with decoder as described above (or if this is bad - hint which way to go) ? 

Thanks and Regards,
rle

jgrandja commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAssertionAuthenticationProvider.java:
```diff
@@ -0,0 +1,90 @@
+ * @author Rafal Lewczuk
+ * @since 0.2.1
+ */
+public final class OAuth2ClientAssertionAuthenticationProvider implements AuthenticationProvider {
```
Please remove this class and integrate the logic in `OAuth2ClientAuthenticationProvider`. With this, `OAuth2PkceCodeVerifier` can be removed as well.

jgrandja commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAssertionAuthenticationProvider.java:
```diff
@@ -0,0 +1,90 @@
+public final class OAuth2ClientAssertionAuthenticationProvider implements AuthenticationProvider {
+	private final RegisteredClientRepository registeredClientRepository;
+	private final OAuth2PkceCodeVerifier pkceCodeVerifier;
+	private final JwtDecoder jwtDecoder;
```
`JwtDecoder` will be required in `OAuth2ClientAuthenticationProvider` but it does not need to be exposed via constructor or setter. The internal implementation will have similarities to [OidcIdTokenDecoderFactory](https://github.com/spring-projects/spring-security/blob/006b9b960797d279b31cf8c8d16f1549c5632b2c/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenDecoderFactory.java#L66).

The logic for selecting key material to validate the JWS is as follows:
1) Lookup `RegisteredClient` using `client_id`
2) If `registeredClient.getClientAuthenticationMethods().contains(ClientAuthenticationMethod.PRIVATE_KEY_JWT)` then obtain `jwks_uri` from the registration and build decoder similar to `OidcIdTokenDecoderFactory`.
3) If `registeredClient.getClientAuthenticationMethods().contains(ClientAuthenticationMethod.CLIENT_SECRET_JWT)` then use  `registeredClient.clientSecret` and build decoder similar to `OidcIdTokenDecoderFactory`.

NOTES:
- `PRIVATE_KEY_JWT` should take priority over `CLIENT_SECRET_JWT`
- `jwks_uri` is defined by [OpenID Connect Dynamic Client Registration 1.0](https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata). This will need to be added to `ClientSettings.getJwkSetUrl()`, which is associated to `RegisteredClient`.

rlewczuk commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/authentication/ClientAssertionAuthenticationConverter.java:
```diff
@@ -0,0 +1,76 @@
+		if (parameters.get(OAuth2ParameterNames.CLIENT_ASSERTION).size() != 1) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_REQUEST));
+		}
+
```
Here things get interesting. Specification (section 3) states that `sub` claim must be the `client_id`, so I was thinking of using it prior to signature verification. Yet it does not cover cases where client assertion is encrypted - such situations would force us to use `kid` header to determine key which can be ambiguous. Is this the reason why we choose to use `client_id` parameter ? Should `sub` claim be used in any way or just verified against `client_id`

rlewczuk replied:
OK, I see... `OAuth2ClientAuthenticationToken` also requires `client_id`, so requiring this parameter makes things simpler.

rlewczuk commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/authentication/ClientAssertionAuthenticationConverter.java:
```diff
@@ -0,0 +1,76 @@
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_REQUEST));
+		}
+
+		Map<String, Object> additionalParameters = new HashMap<>();
```
Method `extractAdditionalParameters(request)` now appears three times with minor change, maybe it's a good candidate to factor out for example to `OAuth2EndpointUtils.extractAdditionalParameters(request, ...excludeParams)` ?

jgrandja replied:
Sure let's share it via `OAuth2EndpointUtils`

rlewczuk commented:
@jgrandja Implemented as per review comments with some remarks below.

I wasn't able to use standard `JwtDecoderFactory` interface in authentication provider to create JWT decoders as two parameters are needed (registered client and authentication method); with only registered client as parameter two token validation methods will exclude themselves (shared secret and private key).

Question regarding validation: RFC states that `aud` and `iss` claims are required. Should we strictly validate those ? Now I'm only checking `iss` if token issuer is configured in registered client settings. Where can I get external server URL (or should audience field also be part of registered client settings) ? 

Sorry for noob questions, I'm still learning logic behind design of Spring projects :) 

Can I ask for review ? 

Regards,
rle

jgrandja commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/authentication/ClientAssertionAuthenticationConverter.java:
```diff
@@ -0,0 +1,91 @@
+		MultiValueMap<String, String> parameters = OAuth2EndpointUtils.getParameters(request);
+
+		// client_id (REQUIRED)
+		String clientId = parameters.getFirst(OAuth2ParameterNames.CLIENT_ID);
```
`client_assertion_type` parameter should be the first check to determine if this `Converter` should be applied. If `client_assertion_type` parameter is not present then return `null` to let the next `Converter` get applied.

jgrandja reviewed (changes requested):
Thanks for the updates @rlewczuk. Please see review comments.

rlewczuk commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/RegisteredClientDecoderFactory.java:
```diff
@@ -0,0 +1,30 @@
+ * @author Rafal Lewczuk
+ * @since 0.2.1
+ */
+public interface RegisteredClientDecoderFactory {
```
I've reimplemented this part using `JwtDecoderFactory<OAuth2AuthenticationContext>` but it seems that `OAuth2ClientAuthenticationToken` would make things cleaner (especially tests) and without need to declare additional context parameter identifiers. I see that `OAuth2ClientAuthenticationToken` has all needed things: `registeredClient` and `clientAuthenticationMethod`. Would it make sense to use it instead ?

jgrandja replied:
> I see that `OAuth2ClientAuthenticationToken` has all needed things: `registeredClient` and `clientAuthenticationMethod`. Would it make sense to use it instead ?

The `OAuth2ClientAuthenticationToken` ("authentication request") is not populated with `registeredClient` initially, it's only populated on the return value ("authentication result") when authentication succeeds. So you would need to pass `OAuth2ClientAuthenticationToken` into `OAuth2AuthenticationContext` and `registeredClient` as a context attribute.

rlewczuk commented:
While we cannot trust in principle JWT tokens prior to verification, I wonder if we can use JWT headers. Digging through nimbus implementation and `NimbusJwetDecoder` I see that it always parses header before verification. It always uses `alg` field in `com.nimbusds.jwt.JWTParser.parse()` method. I wonder if we can do the same in converter to determine exact client authentication method. In certain configurations `kid` fields is also used.

Currently we declare a synthetic client authentication method (in fact twice - in converter and authentication provider in order to not expose it), then we add a bunch of clumsy logic `OAuth2ClientAuthenticationProvider.authenticate` to dispatch it and on top of that we have to work around fact that `JwtDecoder` has single argument and this we pack arguments into context object, which further complicates things (mostly in unit tests, plus we are forced to cram two separate pieces of logic and switch into single decoder factory). 

I'm writing this because - frankly - I suspect that  currently we are adding lot of complexity only to work around non-existent issue. Determining correct authentication method earlier would make things straight and as we already trust `JWTParse.parse()` prior to verification we are free to use it to know exact algorithm early enough (determine authentication method knowing it). 

@jgrandja what do you think ?

rlewczuk commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/RegisteredClientJwtDecoderFactory.java:
```diff
@@ -0,0 +1,137 @@
+		}
+
+		ClientSettings clientSettings = registeredClient.getClientSettings();
+		String jwsAlgorithm = clientSettings.getSetting(ConfigurationSettingNames.Client.TOKEN_SIGNATURE_ALGORIHM);
```
Both token signature algorithm and jwk set url are now implemented in both `ClientSettings`/`OidcClientRegistration` and processed in `OidcRegistrationAuthenticationProvider`. At the moment it only copies those values and I some some form of validation should also be implemented: at minimum that one is proper URL and second is supported algorithm name. But what about higher level validations ? For example checking for missing fields (algorithm) if specific client authentication method is present or looking if both private_key_jwt and client_secret_jwt are enabled at once. Or limiting allowed algorithms to either HMAC or PKI depending on authentication method passed. Should I implement those as well ?

jgrandja replied:
I'm not following the need for "higher level validations". The logic should be this:

- If `ClientSettings.getTokenEndpointSignatureAlgorithm()` is `MacAlgorithm` then use `registeredClient.getClientSecret()` for the secret.
- If `ClientSettings.getTokenEndpointSignatureAlgorithm()` is `SignatureAlgorithm` then use `ClientSettings.getJwkSetUrl()` for the public keys.
- If `ClientSettings.getTokenEndpointSignatureAlgorithm()` is not set, then default to `RS256` and ensure `ClientSettings.getJwkSetUrl()` is set else throw exception.

rlewczuk commented:
We are ready for next round of review. Design and implementation seems to be complete (i.e. won't change without feedback from review), unit tests are here. One missing thing are integration tests. I see some implemented in `configurers.oauth2.server.authorization` tree. Is this particular change suitable for integration tests ? (I'm thinking of register client -> use client scenario)

rlewczuk commented:
Now it's complete as far as I can tell. Please review. Sorry for that chaos :)

jgrandja commented:
Thanks for the updates @rlewczuk.

FYI, I'm on PTO Nov 19-24 (inclusive) and the release is Dec 1, so ideally I'd like to merge this week.
I'll start the review tomorrow and will add a polish commit (if needed) to get it merged this week.
Thanks.

rlewczuk commented:
Ok, I'll be available tomorrow, so if there is still something you want me to do I can react quickly. Now I'll rebase it against latest commits on main branch and push (i.e. no more changes here, only rebase and tests). Thanks.

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/web/authentication/ClientJwtAssertionAuthenticationConverterTests.java:
```diff
@@ -0,0 +1,138 @@
+import static org.assertj.core.api.Assertions.*;
+import static org.assertj.core.api.Assertions.entry;
+
+public class ClientJwtAssertionAuthenticationConverterTests {
```
Rename to `JwtClientAssertionAuthenticationConverterTests`

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/web/authentication/ClientJwtAssertionAuthenticationConverterTests.java:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	@Test
+	public void convertWhenAuthorizationHeaderNotBasicThenReturnNull() {
```
This test is not valid

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/web/authentication/ClientJwtAssertionAuthenticationConverterTests.java:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	@Test
+	public void convertWhenNoClientIdThenNull() {
```
This test should throw `OAuth2AuthenticationException`

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/web/authentication/ClientJwtAssertionAuthenticationConverterTests.java:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	@Test
+	public void convertWhenTwiceClientIdThenThrow() {
```
`..ThenThrow` - Please review naming conventions of test methods in other `AuthenticationConverter` tests and ensure they are consistent

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/TestRegisteredClients.java:
```diff
@@ -64,4 +64,16 @@
 				.clientSettings(ClientSettings.builder().requireProofKey(true).build());
 	}
+
+	public static RegisteredClient.Builder registeredJwtAssertionClient(String alg) {
```
Revert this and reuse existing `registeredClient` or `registeredClient2`

jgrandja reviewed (changes requested):
@rlewczuk I'm not going to be able to merge this before I head out for PTO on Friday as there are a few more changes needed. Also, I haven't finished the detailed review and plan on continuing the review tomorrow with `OAuth2ClientAuthenticationProvider`, `RegisteredClientJwtAssertionDecoderFactory` and `RegisteredClientJwtAssertionValidator`.

After all the comments have been addressed, I should be able to merge when I return.

jgrandja commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/RegisteredClientJwtAssertionValidator.java:
```diff
@@ -0,0 +1,121 @@
+ *
+ * @author Rafal Lewczuk
+ */
+class RegisteredClientJwtAssertionValidator implements OAuth2TokenValidator<Jwt> {
```
Let's try to replace this by reusing `JwtIssuerValidator`, `JwtTimestampValidator`, and `JwtClaimValidator`(s)  and composing in a `DelegatingOAuth2TokenValidator`.

rlewczuk replied:
Heh, this was my initial implementation, then I've changed it after `OidcIdTokenValidator`. I'll revert it in a moment.

jgrandja commented on oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2ClientAuthenticationWithClientJwtAssertionTests.java:
```diff
@@ -0,0 +1,236 @@
+ *
+ * @author Rafal Lewczuk
+ */
+public class OAuth2ClientAuthenticationWithClientJwtAssertionTests {
```
Please remove this and add the following tests in `OidcClientRegistrationTests`:

1) Register client with `client_secret_jwt` and perform `client_credentials` grant to obtain token.
2) Register client with `private_key_jwt` and perform `client_credentials` grant to obtain token.

These tests will ensure full integration is working.

jgrandja reviewed (changes requested):
@rlewczuk I finished the detail review. Please see additional comments.

rlewczuk commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ClientSettings.java:
```diff
@@ -53,6 +53,25 @@ public boolean isRequireAuthorizationConsent() {
 		return getSetting(ConfigurationSettingNames.Client.REQUIRE_AUTHORIZATION_CONSENT);
 	}
 
+	/**
```
Got me thinking when writing those (trivial) tests: if it does make sense for `ClientSettings.getTokenEndpointSigningAlgorithm()` to have default value of `RS256`. I'm implementing it this way, let me know if this is wrong.

rlewczuk commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/OidcClientMetadataClaimAccessor.java:
```diff
@@ -155,4 +155,23 @@ default URL getRegistrationClientUrl() {
+	 * @return {@code URL} of JWK Set used by Client to sign requests to Server
+	 * @since 0.2.1
+	 */
+	default String getJwksUri() {
```
Changed this way but now we have `URL getJwkSetUrl()` and in builder: `Builder jwkSetUrl(String)`. Is it valid ? Or should we go for symmetry here ?

rlewczuk commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/RegisteredClientJwtAssertionDecoderFactory.java:
```diff
@@ -0,0 +1,157 @@
+		}
+		boolean isHmac = JCA_ALGORITHM_MAPPINGS.containsKey(signatureAlgorithm);
+
+		// No fancy .computerIfAbsent() calls as we need to check if client configuration has changed
```
I've implement `buildDecoder()` to be (almost) identical as in `OidcIdTokenDecoderFactory` (error messages differ), but now it has less guards against  client misconfiguration. Missing checks are done when client registers, so it applies only to admin/operator misconfiguration.

It still checks if registered client configuration has changed but it's now simplified to this: `cachedDecoder != null && registeredClient.equals(cachedDecoder.registeredClient)`. I'd rather stick to it if possible, otherwise it means maintenance problems for those who configure static clients and periodically rotate keys.

Overall we have around 25 LOC less (plus validator class removed) - not much but from logic-dense regions.

rlewczuk commented on oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAuthenticationProvider.java:
```diff
@@ -100,19 +121,41 @@ public Authentication authenticate(Authentication authentication) throws Authent
+			throwInvalidClient("authentication_method");
+		}
+
+		if (CLIENT_ASSERTION_AUTHENTICATION_METHOD.equals(clientAuthentiationMethod) && this.decoderFactory == null) {
```
I see that one of last commits makes `ProviderSettings` bean optional and it has been removed from samples. I would either revert those removal (probably bad idea) or try to extract token endpoint URL in similar way issuer URL is resolved currently. I'll try to implement this latter approach in token endpoint filter, please protest if this is wrong approach.

rlewczuk commented:
@jgrandja all changes from review are done plus I discovered some missing parts in oauth2 discovery and oidc discovery endpoints. Amount of changes grew, it becomes a monster patch of more than 2000 LOC. 

One of recent changes for resolving issuer from current request did wreak a bit of havoc: I've resigned from using statically configured issuer URL from `ProviderSettings` and using value resolved in `OAuth2AuthorizationServerMetadataEndpointFilter`. This had (potential) impact on `RegisteredClientJwtAssertionDecoderFactory` which would require not only `RegisteredClient` but also proper issuer URL. I decided against changing it as it caused unnecessary complications in both caching and subclassing `OAuth2AuthenticationContext`. Instead I'm verifying `aud` claim directly in `OAuth2ClientAuthenticationProvider`. For reference implementation with `aud` verification in decoder is here: https://github.com/rlewczuk/spring-authorization-server/tree/jwt-client-auth-aud-verifier. Also, I'm verifying URI against endpoint client is calling (`/oauth2/token`, `/oauth2/introspect` or `/oauth2/revoke`). I didn't find anything in RFCs clarifying this, so I'm guessing `aud` should point to appropriate endpoint, not just blindly on `/oauth2/token`. 

I'm marking PR as ready for review. I'll probably also look myself at it in coming days, so some small patches/cleanups/fixes/tests may appear but it won't cause any major changes. 

Thanks for passing those fixes to me, not just proceeding with polishing commit. In terms of climbing up the learning curve that last round was probably the biggest uplift for me.

Note: I'll be absent on Monday, 29th, will be available on all remaining days this month.

Thanks and Regards,
rle

jgrandja commented:
Thanks for all the updates @rlewczuk.

We're definitely getting close but we'll have to continue this after `0.2.1` is released.
I don't want to rush this as there are quite a few changes.

Also, I'm thinking about breaking up this PR into smaller PR's so it's easier to consume and more efficient to review.
But please hold off for now until I come up with a plan.

I'll get back to you later next week after the release is out.

Thanks for all your work!

rlewczuk commented:
Yep, at some point I also started thinking if one big commit is valid approach. Split into authentication (on preconfigured client), OIDC registration and discovery parts seems to be natural.

rlewczuk commented:
Hi @jgrandja. I've rebased code onto newly released version. Now one thing has to be decided: do we split this PR into parts or try to proceed as it is. As we are at the beginning of iteration, there is time to fix flaws in subsequent PRs (granted that there are no serious ones).

jgrandja commented:
@rlewczuk We're going to continue with this PR as we're pretty close and it's not needed to split up at this point.

Before I continue with next review, please apply these changes:

1) Remove `token_endpoint_auth_signing_alg_values_supported`, `revocation_endpoint_auth_signing_alg_values_supported` and `introspection_endpoint_auth_signing_alg_values_supported`. These provider configuration metadata are OPTIONAL and is not needed in this PR. We can consider adding it in a subsequent PR.
2) Make `RegisteredClientJwtAssertionDecoderFactory` as an inner `private static class` in `OAuth2ClientAuthenticationProvider`.

I should be available to review within the next couple of days. Thank you.

rlewczuk commented:
Both changes are applied.

jgrandja commented:
Thanks for all the updates @rlewczuk.

After reviewing, I found a few more changes were required (especially in tests) and my goal was to have this merged before end of week, as I'm off for the following 2 weeks. So I decided to add a polish commit to get this merged.

Please let me know if you have any questions in regards to the changes applied in the polish commit.

Thanks again for all your time in contributing to this very important feature !

rlewczuk commented:
Quite extensive changes, I have some questions (of "how things work here" kind) but I need to dig into this polish commit a bit deeper first.

I have two general questions:

* I've set aside patch with additional server metadata as per your previous request; can I post PR with this change now ? 

* Can I start work on client assertion grants (second part of this RFC) ? 

Regards,
rle

rlewczuk commented:
Regarding `JwtClientAssertionDecoderFactory` caching: I see that cache refresh logic has been removed and replaced with simple `.computeIfAbsent()` call. What is rationale behind it ? Is it assumed that client registrations are immutable once created ?

jgrandja commented:
@rlewczuk 

> I've set aside patch with additional server metadata as per your previous request; can I post PR with this change now ?

The patch would not work as you specified all the JWA algorithms and that would not be the case for all deployments. It depends what keys are configured in `JWKSource<SecurityContext>` per deployment and likely it would be a very small set of the superset. This would differ is each deployment. Given this is OPTIONAL and IMO is not really needed, we'll leave it out for now.

> Can I start work on client assertion grants (second part of this RFC) ?

Please log a new issue and we'll see if this is a priority issue based on upvotes. There are other features that are higher in priority compared to this one so we'll leave this on hold for now.

> Regarding `JwtClientAssertionDecoderFactory` caching...

I consider the "refresh cache" logic a potential future enhancement. It's not needed at this time and I'd rather keep things as simple as possible with the initial implementation. We might introduce this at a later point if it becomes an issue.
